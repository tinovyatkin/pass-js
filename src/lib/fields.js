'use strict';

/**
 * Field accessors class
 */

const only = require('./only');
const { getW3CDateString } = require('./w3cdate');

class Fields {
  /**
   * Creates an instance of Fields
   *
   * @param {import('../pass')} pass - parent Pass class
   * @param {string} key - key name that these fields are bound to
   * @memberof Fields
   */
  constructor(pass, key) {
    this.pass = pass;
    this.key = key;
    Object.preventExtensions(this);
  }

  /**
   * Returns an array of all fields.
   *
   * @returns {{key: string, value: string, label: string}[]}
   * @memberof Fields
   */
  all() {
    if (!(this.key in this.pass.structure)) this.pass.structure[this.key] = [];
    return this.pass.structure[this.key];
  }

  /**
   * Adds a field to the end of the list
   *
   * @param {string | {key: string, label: string, value: string}} key - Field key or object with all fields
   * @param {string} [label] - Field label (optional)
   * @param {string} [value] - Field value
   * @param {Object} [options] - Other field options (e.g. dateStyle)
   * @returns {Fields}
   * @memberof Fields
   */
  add(key, label, value, options) {
    if (arguments.length > 1) {
      this.remove(key);
      const field = { key, value };
      if (label) field.label = label;
      if (options) Object.assign(field, options);
      this.all().push(field);
    } else if (Array.isArray(key)) {
      for (const k of key) this.add(k);
    } else {
      this.remove(key.key);
      // save object copy
      this.all().push(Object.assign({}, key));
    }
    return this;
  }

  /**
   * Returns a field
   *
   * @param {string} key
   * @returns {{key: string, label: string, value: string}} If field exists, returns an object with common keys and rest of keys
   * @memberof Fields
   */
  get(key) {
    return this.all().find(v => v.key === key);
  }

  /**
   * Sets value field for a given key
   *
   * @param {string} key
   * @param {string} value
   * @memberof Fields
   */
  setValue(key, value) {
    if (typeof key !== 'string')
      throw new TypeError(
        `key for setValue must be a string, received ${typeof key}`,
      );
    if (typeof value !== 'string')
      throw new TypeError(
        `value for setValue must be a string, received ${typeof value}`,
      );
    const field = this.get(key);
    if (!field) return this.add({ key, value });
    field.value = value;
    return this;
  }

  /**
   * Removes a given field.
   *
   * @param {string} key
   * @returns {Fields}
   * @memberof Fields
   */
  remove(key) {
    const idx = this.all().findIndex(v => v.key === key);
    if (idx > -1) {
      this.all().splice(idx, 1);
      // remove property completely if there is no fields left
      if (this.all().length === 0) this.clear();
    }
    return this;
  }

  /**
   * Set a field as Date value with appropriated options
   *
   * @param {string} key
   * @param {string} label
   * @param {Date} date
   * @param {{dateStyle?: string, ignoresTimeZone?: boolean, isRelative?: boolean, timeStyle?:string}} [formatOptions]
   * @returns {Fields}
   * @throws if date is not a Date or invalid Date
   * @memberof Fields
   */
  setDateTime(key, label, date, formatOptions = {}) {
    if (!(date instanceof Date) || !isFinite(date))
      throw new Error(
        'First parameter of setDateTime must be an instance of Date',
      );
    //  Either specify both a date style and a time style, or neither.
    if (
      (!('dateStyle' in formatOptions) && 'timeStyle' in formatOptions) ||
      ('dateStyle' in formatOptions && !('timeStyle' in formatOptions))
    )
      throw new Error(
        'Either specify both a date style and a time style, or neither',
      );
    // adding
    this.add(
      Object.assign(
        {
          key,
          label,
          value: getW3CDateString(date),
        },
        only(
          formatOptions,
          'dateStyle timeStyle ignoresTimeZone isRelative changeMessage',
        ),
      ),
    );
    return this;
  }

  /**
   * Removes all fields.
   *
   * @returns {Fields}
   * @memberof Fields
   */
  clear() {
    delete this.pass.structure[this.key];
    return this;
  }
}

module.exports = Fields;
