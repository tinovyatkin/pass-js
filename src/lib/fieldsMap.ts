'use strict';

import { Field, FieldDescriptor, DataStyleFormat } from '../interfaces';

import { getW3CDateString } from './w3cdate';

export class FieldsMap extends Map<string, FieldDescriptor> {
  /**
   * Returns Map as array of objects with key moved inside like a key property
   */
  toJSON(): Field[] | undefined {
    if (!this.size) return undefined;
    return [...this].map(
      ([key, data]): Field => {
        // Remap Date objects to string
        if (data.value instanceof Date)
          data.value = getW3CDateString(data.value);
        return { key, ...data };
      },
    );
  }

  /**
   * Adds a field to the end of the list
   *
   * @param {Field} field - Field key or object with all fields
   * @returns {FieldsMap}
   * @memberof FieldsMap
   */
  add(field: Field): FieldsMap {
    const { key, ...data } = field;
    if (typeof key !== 'string')
      throw new TypeError(
        `To add a field you must provide string key value, received ${typeof key}`,
      );
    if (!('value' in data))
      throw new TypeError(
        `To add a field you must provide a value field, received: ${JSON.stringify(
          data,
        )}`,
      );
    this.set(key, data);
    return this;
  }

  /**
   * Sets value field for a given key, without changing the rest of field properties
   *
   * @param {string} key
   * @param {string} value
   * @memberof FieldsMap
   */
  setValue(key: string, value: string): FieldsMap {
    if (typeof key !== 'string')
      throw new TypeError(
        `key for setValue must be a string, received ${typeof key}`,
      );
    if (typeof value !== 'string')
      throw new TypeError(
        `value for setValue must be a string, received ${typeof value}`,
      );
    const field = this.get(key) || { value };
    field.value = value;
    this.set(key, field);
    return this;
  }

  /**
   * Set a field as Date value with appropriated options
   *
   * @param {string} key
   * @param {string} label
   * @param {Date} date
   * @param {{dateStyle?: string, ignoresTimeZone?: boolean, isRelative?: boolean, timeStyle?:string}} [formatOptions]
   * @returns {FieldsMap}
   * @throws if date is not a Date or invalid Date
   * @memberof FieldsMap
   */
  setDateTime(
    key: string,
    label: string,
    date: Date,
    {
      dateStyle,
      ignoresTimeZone,
      isRelative,
      timeStyle,
    }: {
      dateStyle?: DataStyleFormat;
      ignoresTimeZone?: boolean;
      isRelative?: boolean;
      timeStyle?: DataStyleFormat;
    } = {},
  ): FieldsMap {
    if (typeof key !== 'string')
      throw new TypeError(`Key must be a string, received ${typeof key}`);
    if (typeof label !== 'string')
      throw new TypeError(`Label must be a string, received ${typeof label}`);
    if (!(date instanceof Date))
      throw new TypeError(
        'Third parameter of setDateTime must be an instance of Date',
      );
    //  Either specify both a date style and a time style, or neither.
    if (!!dateStyle !== !!timeStyle)
      throw new ReferenceError(
        'Either specify both a date style and a time style, or neither',
      );
    // adding
    this.set(key, {
      label,
      value: date,
      dateStyle,
      ignoresTimeZone,
      isRelative,
      timeStyle,
    });

    return this;
  }
}
