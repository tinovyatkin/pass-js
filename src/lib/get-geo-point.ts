/**
 * Returns normalized geo point object from geoJSON, {lat, lng} or {lattitude,longutude,altitude}
 *
 * @param {number[] | { lat: number, lng: number, alt?: number } | { longitude: number, latitude: number, altitude?: number }} point
 * @returns {{ longitude: number, latitude: number, altitude?: number }}
 * @throws on unknown point format
 */
export function getGeoPoint(
  point:
    | number[]
    | { lat: number; lng: number; alt?: number }
    | { longitude: number; latitude: number } & (
        | { altitude?: number }
        | { elevation?: number }),
): { longitude: number; latitude: number; altitude?: number } {
  if (!point) throw new Error("Can't get coordinates from undefined");

  // GeoJSON Array [longitude, latitude(, elevation)]
  if (Array.isArray(point)) {
    if (point.length < 2 || !point.every(n => Number.isFinite(n)))
      throw new Error(
        `Invalid GeoJSON array of numbers, length must be 2 to 3, received ${
          point.length
        }`,
      );
    return {
      longitude: point[0],
      latitude: point[1],
      altitude: point[2],
    };
  }

  // it can be an object with both lat and lng properties
  if ('lat' in point && 'lng' in point) {
    return {
      longitude: point.lng,
      latitude: point.lat,
      altitude: point.alt,
    };
  }

  if ('longitude' in point && 'latitude' in point) {
    // returning a copy
    return {
      longitude: point.longitude,
      latitude: point.latitude,
      altitude:
        'altitude' in point
          ? point.altitude
          : 'elevation' in point
          ? point.elevation
          : undefined,
    };
  }

  // If we are here it means we can't understand what a hell is it
  throw new Error(`Unknown geo point format: ${JSON.stringify(point)}`);
}
