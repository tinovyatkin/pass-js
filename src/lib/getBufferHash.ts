'use strict';

import { createHash } from 'crypto';

/**
 *
 * @param {Buffer} buffer
 * @returns {string}
 */
export function getBufferHash(buffer: Buffer | string): string {
  // creating hash
  const sha = createHash('sha1');
  sha.update(buffer);
  return sha.digest('hex');
}
