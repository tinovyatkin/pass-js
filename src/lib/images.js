/**
 * Base PassImages class to add image filePath manipulation
 */

'use strict';

const { basename, extname, resolve } = require('path');
const { stat, readdir } = require('fs').promises;

// Supported images.
const { IMAGES, DENSITIES } = require('../constants');

class PassImages {
  constructor() {
    // Creating this way to make it invisible
    this.map = new Map();

    // define setters and getters for particular images
    for (const imageType in IMAGES) {
      Object.defineProperty(this, imageType, {
        enumerable: false,
        get: this.getImage.bind(this, imageType),
        set: this.setImage.bind(this, imageType, '1x'),
      });
      // setting retina properties too
      for (const density of DENSITIES) {
        Object.defineProperty(this, imageType + density, {
          enumerable: false,
          get: this.getImage.bind(this, imageType, density),
          set: this.setImage.bind(this, imageType, density),
        });
      }
    }

    Object.preventExtensions(this);
  }

  /**
   * Returns a given imageType path with a density
   *
   * @param {string} imageType
   * @param {string} density - can be '2x' or '3x'
   * @returns {string} - image path
   * @memberof PassImages
   */
  getImage(imageType, density = '1x') {
    if (!(imageType in IMAGES))
      throw new Error(`Requested unknown image type: ${imageType}`);
    if (!DENSITIES.includes(density))
      throw new Error(`Invalid desity for "${imageType}": ${density}`);
    if (!this.map.has(imageType)) return undefined;
    return this.map.get(imageType).get(density);
  }

  /**
   * Saves a given imageType path
   *
   * @param {string} imageType
   * @param {string} density
   * @param {string} fileName
   * @memberof PassImages
   */
  setImage(imageType, density = '1x', fileName) {
    if (!(imageType in IMAGES))
      throw new Error(`Attempted to set unknown image type: ${imageType}`);
    const imgData = this.map.get(imageType) || new Map();
    imgData.set(density, fileName);
    this.map.set(imageType, imgData);
  }

  /**
   * Load all images from the specified directory. Only supported images are
   * loaded, nothing bad happens if directory contains other files.
   *
   * @param {string} dir - path to a directory with images
   * @memberof PassImages
   */
  async loadFromDirectory(dir) {
    const fullPath = resolve(dir);
    const stats = await stat(fullPath);
    if (!stats.isDirectory())
      throw new Error(`Path ${fullPath} must be a directory!`);

    const files = await readdir(fullPath);
    for (const filePath of files) {
      // we are interesting only in PNG files
      if (extname(filePath) === '.png') {
        const fileName = basename(filePath, '.png');
        // this will split imagename like background@2x into 'background' and '2x' and fail on anything else
        const [, imageType, , density] =
          /^([a-z]+)(@([23]x))?$/.exec(fileName) || [];
        if (imageType in IMAGES && (!density || DENSITIES.includes(density)))
          this.setImage(imageType, density, resolve(fullPath, filePath));
      }
    }

    return this;
  }
}

module.exports = PassImages;
