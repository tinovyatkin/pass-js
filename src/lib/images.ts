/**
 * Base PassImages class to add image filePath manipulation
 */

'use strict';

import { promisify } from 'util';
import * as path from 'path';
import { createReadStream, promises as fs } from 'fs';

import glob from 'fast-glob';
import * as imagesize from 'imagesize';

import { IMAGES, DENSITIES } from '../constants';

interface ImageSizeResult {
  format: 'gif' | 'png' | 'jpeg';
  width: number;
  height: number;
}

const imageSize: (
  v: import('stream').Readable,
) => Promise<ImageSizeResult> = promisify(imagesize);

type ImageDensity = '1x' | '2x' | '3x';
type ImageType =
  | 'logo'
  | 'icon'
  | 'background'
  | 'footer'
  | 'strip'
  | 'thumbnail';

const IMAGES_GLOB = `(${Object.keys(IMAGES).join('|')})*(@2x|@3x).png`;
const IMAGES_TYPES = new Set(Object.keys(IMAGES));

export class PassImages {
  private map: Map<string, string | Buffer> = new Map();

  constructor(images?: PassImages) {
    if (images instanceof PassImages) {
      this.map = new Map([...images.map]);
    }
    Object.preventExtensions(this);
  }

  get count(): number {
    return this.map.size;
  }

  async toArray(): Promise<{ path: string; data: Buffer }[]> {
    return Promise.all(
      [...this.map].map(async ([filepath, pathOrBuffer]) => ({
        path: filepath,
        data:
          typeof pathOrBuffer === 'string'
            ? await fs.readFile(pathOrBuffer)
            : pathOrBuffer,
      })),
    );
  }

  /**
   * Checks that all required images is set or throws elsewhere
   */
  validate(): void {
    const keys = [...this.map.keys()];
    // Check for required images
    for (const requiredImage of ['icon', 'logo'])
      if (!keys.some(img => img.endsWith(`${requiredImage}.png`)))
        throw new SyntaxError(`Missing required image ${requiredImage}.png`);
  }

  /**
   * Load all images from the specified directory. Only supported images are
   * loaded, nothing bad happens if directory contains other files.
   *
   * @param {string} dir - path to a directory with images
   * @memberof PassImages
   */
  async loadFromDirectory(dir: string): Promise<PassImages> {
    for await (const file of glob.stream(
      [path.join(dir, IMAGES_GLOB), path.join(dir, '*.lproj', IMAGES_GLOB)],
      {
        onlyFiles: true,
        deep: 1,
      },
    )) {
      if (typeof file !== 'string') continue;
      // getting image type and optional density and language
      const re = /((?<lang>[-A-Z_a-z]+)\.lproj\/)?(?<imageType>[a-z]+)(@(?<density>[23]x))?\.png$/.exec(
        file,
      );
      if (!re) continue;
      const { imageType, density, lang } = re.groups as {
        imageType: ImageType;
        density?: ImageDensity;
        lang?: string;
      };
      await this.setImage(imageType, file, density, lang);
    }

    return this;
  }

  async setImage(
    imageType: ImageType,
    pathOrBuffer: string | Buffer,
    density?: ImageDensity,
    lang?: string,
  ): Promise<void> {
    if (!IMAGES_TYPES.has(imageType))
      throw new TypeError(`Unknown image type ${imageSize} for ${imageType}`);
    if (density && !DENSITIES.has(density))
      throw new TypeError(`Invalid density ${density} for ${imageType}`);

    // check data
    let sizeRes;
    if (typeof pathOrBuffer === 'string') {
      const rs = createReadStream(pathOrBuffer);
      sizeRes = await imageSize(rs);
      rs.destroy();
    } else {
      if (!Buffer.isBuffer(pathOrBuffer))
        throw new TypeError(
          `Image data for ${imageType} must be either file path or buffer`,
        );
      const { Parser } = imagesize;
      const parser = Parser();
      const res = parser.parse(pathOrBuffer);
      if (!res !== Parser.DONE)
        throw new TypeError(
          `Supplied buffer doesn't contain valid PNG image for ${imageType}`,
        );
      sizeRes = parser.getResult() as ImageSizeResult;
    }
    this.checkImage(imageType, sizeRes, density);
    this.map.set(this.getImageFilename(imageType, density, lang), pathOrBuffer);
  }

  // eslint-disable-next-line complexity
  private checkImage(
    imageType: ImageType,
    sizeResult: ImageSizeResult,
    density?: ImageDensity,
  ): void {
    const densityMulti = density ? parseInt(density.charAt(0), 10) : 1;
    const { format, width, height } = sizeResult;
    if (format !== 'png')
      throw new TypeError(`Image for "${imageType}" is not a PNG file!`);
    if (!Number.isInteger(width) || width <= 0)
      throw new TypeError(`Image ${imageType} has invalid width: ${width}`);
    if (!Number.isInteger(height) || height <= 0)
      throw new TypeError(`Image ${imageType} has invalid height: ${height}`);
    /**
     * @see {@link https://developer.apple.com/library/archive/documentation/UserExperience/Conceptual/PassKit_PG/Creating.html}
     */
    switch (imageType) {
      case 'icon':
        if (width !== 29 * densityMulti)
          throw new TypeError(
            `icon image must have width ${29 *
              densityMulti}px for ${densityMulti}x density`,
          );
        if (height !== 29 * densityMulti)
          throw new TypeError(
            `icon image must have height ${29 *
              densityMulti}px for ${densityMulti}x density`,
          );
        break;

      case 'logo':
        if (width > 160 * densityMulti)
          throw new TypeError(
            `logo image must have width no large than ${160 *
              densityMulti}px for ${densityMulti}x density`,
          );
        // if (height > 50 * densityMulti)
        //   throw new TypeError(
        //     `logo image must have height ${50 *
        //       densityMulti}px for ${densityMulti}x density, received ${height}`,
        //   );
        break;

      case 'background':
        if (width > 180 * densityMulti)
          throw new TypeError(
            `background image must have width ${180 *
              densityMulti}px for ${densityMulti}x density`,
          );
        if (height > 220 * densityMulti)
          throw new TypeError(
            `background image must have height ${220 *
              densityMulti}px for ${densityMulti}x density`,
          );
        break;

      case 'footer':
        if (width > 286 * densityMulti)
          throw new TypeError(
            `footer image must have width ${286 *
              densityMulti}px for ${densityMulti}x density`,
          );
        if (height > 15 * densityMulti)
          throw new TypeError(
            `footer image must have height ${15 *
              densityMulti}px for ${densityMulti}x density`,
          );
        break;

      case 'strip':
        // if (width > 375 * densityMulti)
        //   throw new TypeError(
        //     `strip image must have width ${375 *
        //       densityMulti}px for ${densityMulti}x density, received ${width}`,
        //   );
        if (height > 144 * densityMulti)
          throw new TypeError(
            `strip image must have height ${144 *
              densityMulti}px for ${densityMulti}x density`,
          );
        break;

      case 'thumbnail':
        if (width > 120 * densityMulti)
          throw new TypeError(
            `thumbnail image must have width no large than ${90 *
              densityMulti}px for ${densityMulti}x density, received ${width}`,
          );
        if (height > 150 * densityMulti)
          throw new TypeError(
            `thumbnail image must have height ${90 *
              densityMulti}px for ${densityMulti}x density, received ${height}`,
          );
        break;
    }
  }

  private getImageFilename(
    imageType: ImageType,
    density?: ImageDensity,
    lang?: string,
  ): string {
    return `${lang ? `${lang}.lproj/` : ''}${imageType}${
      /^[23]x$/.test(density || '') ? `@${density}` : ''
    }.png`;
  }
}
