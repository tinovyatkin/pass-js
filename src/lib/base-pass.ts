import { ApplePass } from '../interfaces';
import { BARCODES_FORMAT, STRUCTURE_FIELDS } from '../constants';

import { PassColor } from './pass-color';
import { PassImages } from './images';
import { Localizations } from './localizations';
import { getGeoPoint } from './get-geo-point';
import { PassStructure } from './pass-structure';

const STRUCTURE_FIELDS_SET = new Set([...STRUCTURE_FIELDS, 'nfc']);

export class PassBase extends PassStructure {
  images: PassImages;
  protected localization = new Localizations();

  constructor(fields: Partial<ApplePass> = {}, images?: PassImages) {
    super(fields);

    // restore via setters
    for (const [key, value] of Object.entries(fields)) {
      if (!STRUCTURE_FIELDS_SET.has(key) && key in this) {
        this[key] = value;
      }
    }

    // copy images
    this.images = new PassImages(images);
  }

  // Returns the pass.json object (not a string).
  toJSON(): Partial<ApplePass> {
    return { ...this.fields, formatVersion: 1 };
  }

  get passTypeIdentifier(): string | undefined {
    return this.fields.passTypeIdentifier;
  }

  set passTypeIdentifier(v: string | undefined) {
    if (!v) delete this.fields.passTypeIdentifier;
    else this.fields.passTypeIdentifier = v;
  }

  get teamIdentifier(): string | undefined {
    return this.fields.teamIdentifier;
  }

  set teamIdentifier(v: string | undefined) {
    if (!v) delete this.fields.teamIdentifier;
    else this.fields.teamIdentifier = v;
  }

  get serialNumber(): string | undefined {
    return this.fields.serialNumber;
  }

  set serialNumber(v: string | undefined) {
    if (!v) delete this.fields.serialNumber;
    else this.fields.serialNumber = v;
  }

  /**
   *  Indicates that the pass is void—for example, a one time use coupon that has been redeemed.
   *
   * @type {boolean}
   */
  get voided(): boolean {
    return !!this.fields.voided;
  }

  set voided(v: boolean) {
    if (v) this.fields.voided = true;
    else delete this.fields.voided;
  }

  /**
   * Date and time when the pass expires.
   *
   */
  get expirationDate(): ApplePass['expirationDate'] {
    if (typeof this.fields.expirationDate === 'string')
      return new Date(this.fields.expirationDate);
    return this.fields.expirationDate;
  }
  set expirationDate(v: ApplePass['expirationDate']) {
    if (!v) delete this.fields.expirationDate;
    else {
      if (typeof v === 'string' || v instanceof Date)
        this.fields.expirationDate = v;
      else
        throw new TypeError(
          `expirationDate should be either string or Date, received ${typeof v}`,
        );
    }
  }

  /**
   * Date and time when the pass becomes relevant. For example, the start time of a movie.
   * Recommended for event tickets and boarding passes; otherwise optional.
   *
   * @type {string | Date}
   */
  get relevantDate(): ApplePass['relevantDate'] {
    if (typeof this.fields.relevantDate === 'string')
      return new Date(this.fields.relevantDate);
    return this.fields.relevantDate;
  }

  set relevantDate(v: ApplePass['relevantDate']) {
    if (!v) delete this.fields.relevantDate;
    else {
      if (typeof v === 'string' || v instanceof Date)
        this.fields.relevantDate = v;
      else
        throw new TypeError(
          `relevantDate should be either string or Date, received ${typeof v}`,
        );
    }
  }

  /**
   * A list of iTunes Store item identifiers for the associated apps.
   * Only one item in the list is used—the first item identifier for an app
   * compatible with the current device.
   * If the app is not installed, the link opens the App Store and shows the app.
   * If the app is already installed, the link launches the app.
   */
  get associatedStoreIdentifiers(): ApplePass['associatedStoreIdentifiers'] {
    return this.fields.associatedStoreIdentifiers;
  }
  set associatedStoreIdentifiers(v: ApplePass['associatedStoreIdentifiers']) {
    if (!v) {
      delete this.fields.associatedStoreIdentifiers;
      return;
    }
    const arrayOfNumbers = v.filter(n => Number.isInteger(n));
    if (arrayOfNumbers.length > 0)
      this.fields.associatedStoreIdentifiers = arrayOfNumbers;
    else delete this.fields.associatedStoreIdentifiers;
  }

  /**
   * Brief description of the pass, used by the iOS accessibility technologies.
   * Don’t try to include all of the data on the pass in its description,
   * just include enough detail to distinguish passes of the same type.
   */
  get description(): string | undefined {
    return this.fields.description;
  }

  set description(v: string | undefined) {
    if (!v) delete this.fields.description;
    else this.fields.description = v;
  }

  /**
   * Display name of the organization that originated and signed the pass.
   */
  get organizationName(): string | undefined {
    return this.fields.organizationName;
  }
  set organizationName(v: string | undefined) {
    if (!v) delete this.fields.organizationName;
    else this.fields.organizationName = v;
  }

  /**
   * Optional for event tickets and boarding passes; otherwise not allowed.
   * Identifier used to group related passes.
   * If a grouping identifier is specified, passes with the same style,
   * pass type identifier, and grouping identifier are displayed as a group.
   * Otherwise, passes are grouped automatically.
   * Use this to group passes that are tightly related,
   * such as the boarding passes for different connections of the same trip.
   */
  get groupingIdentifier(): string | undefined {
    return this.fields.groupingIdentifier;
  }

  set groupingIdentifier(v: string | undefined) {
    if (!v) delete this.fields.groupingIdentifier;
    else this.fields.groupingIdentifier = v;
  }

  /**
   * If true, the strip image is displayed without a shine effect.
   * The default value prior to iOS 7.0 is false.
   * In iOS 7.0, a shine effect is never applied, and this key is deprecated.
   */
  get suppressStripShine(): boolean {
    return !!this.fields.suppressStripShine;
  }
  set suppressStripShine(v: boolean) {
    if (!v) delete this.fields.suppressStripShine;
    else this.fields.suppressStripShine = true;
  }

  /**
   * Text displayed next to the logo on the pass.
   */
  get logoText(): string | undefined {
    return this.fields.logoText;
  }
  set logoText(v: string | undefined) {
    if (!v) delete this.fields.logoText;
    else this.fields.logoText = v;
  }

  /**
   * The URL of a web service that conforms to the API described in PassKit Web Service Reference.
   * The web service must use the HTTPS protocol; the leading https:// is included in the value of this key.
   * On devices configured for development, there is UI in Settings to allow HTTP web services.
   *
   * @see {@link https://developer.apple.com/library/archive/documentation/PassKit/Reference/PassKit_WebService/WebService.html#//apple_ref/doc/uid/TP40011988}
   */
  get webServiceURL(): URL | string | undefined {
    return this.fields.webServiceURL;
  }
  set webServiceURL(v: URL | string | undefined) {
    if (!v) {
      delete this.fields.webServiceURL;
      return;
    }

    // validating URL, it will throw on bad value
    const url = v instanceof URL ? v : new URL(v);
    if (url.protocol !== 'https:')
      throw new TypeError(`webServiceURL must be on HTTPS!`);
    this.fields.webServiceURL = url;
  }

  /**
   * The authentication token to use with the web service.
   * The token must be 16 characters or longer.
   */
  get authenticationToken(): string | undefined {
    return this.fields.authenticationToken;
  }
  set authenticationToken(v: string | undefined) {
    if (!v) {
      delete this.fields.authenticationToken;
      return;
    }

    if (typeof v !== 'string')
      throw new TypeError(
        `authenticationToken must be a string, received ${typeof v}`,
      );
    if (v.length < 16)
      throw new TypeError(
        `authenticationToken must must be 16 characters or longer`,
      );
    this.fields.authenticationToken = v;
  }

  /**
   * Background color of the pass, specified as an CSS-style RGB triple.
   *
   * @example rgb(23, 187, 82)
   */
  get backgroundColor():
    | [number, number, number]
    | string
    | undefined
    | PassColor {
    if (!(this.fields.backgroundColor instanceof PassColor)) return undefined;
    return this.fields.backgroundColor.getRGB();
  }
  set backgroundColor(
    v: string | [number, number, number] | undefined | PassColor,
  ) {
    if (!v) {
      delete this.fields.backgroundColor;
      return;
    }
    if (!(this.fields.backgroundColor instanceof PassColor))
      this.fields.backgroundColor = new PassColor(v);
    else this.fields.backgroundColor.set(v);
  }

  /**
   * Foreground color of the pass, specified as a CSS-style RGB triple.
   *
   * @example rgb(100, 10, 110)
   */
  get foregroundColor():
    | [number, number, number]
    | string
    | undefined
    | PassColor {
    if (!(this.fields.foregroundColor instanceof PassColor)) return undefined;
    return this.fields.foregroundColor.getRGB();
  }
  set foregroundColor(
    v: string | [number, number, number] | PassColor | undefined,
  ) {
    if (!v) {
      delete this.fields.foregroundColor;
      return;
    }
    if (!(this.fields.foregroundColor instanceof PassColor))
      this.fields.foregroundColor = new PassColor(v);
    else this.fields.foregroundColor.set(v);
  }

  /**
   * Color of the label text, specified as a CSS-style RGB triple.
   *
   * @example rgb(255, 255, 255)
   */
  get labelColor(): [number, number, number] | string | undefined | PassColor {
    if (!(this.fields.labelColor instanceof PassColor)) return undefined;
    return this.fields.labelColor.getRGB();
  }
  set labelColor(v: string | [number, number, number] | PassColor | undefined) {
    if (!v) {
      delete this.fields.labelColor;
      return;
    }
    if (!(this.fields.labelColor instanceof PassColor))
      this.fields.labelColor = new PassColor(v);
    else this.fields.labelColor.set(v);
  }

  /**
   * Maximum distance in meters from a relevant latitude and longitude that the pass is relevant.
   * This number is compared to the pass’s default distance and the smaller value is used.
   */
  get maxDistance(): number | undefined {
    return this.fields.maxDistance;
  }
  set maxDistance(v: number | undefined) {
    if (!v) {
      delete this.fields.maxDistance;
      return;
    }
    if (!Number.isInteger)
      throw new TypeError(
        'maxDistance must be a positive integer distance in meters!',
      );
    this.fields.maxDistance = v;
  }

  /**
   * Beacons marking locations where the pass is relevant.
   */
  get beacons(): ApplePass['beacons'] {
    return this.fields.beacons;
  }
  set beacons(v: ApplePass['beacons']) {
    if (!v || !Array.isArray(v)) {
      delete this.fields.beacons;
      return;
    }
    for (const beacon of v) {
      if (!beacon.proximityUUID)
        throw new TypeError(`each beacon must contain proximityUUID`);
    }
    // copy array
    this.fields.beacons = [...v];
  }

  /**
   * Information specific to the pass’s barcode.
   * The system uses the first valid barcode dictionary in the array.
   * Additional dictionaries can be added as fallbacks.
   */
  get barcodes(): ApplePass['barcodes'] {
    return this.fields.barcodes;
  }
  set barcodes(v: ApplePass['barcodes']) {
    if (!v) {
      delete this.fields.barcodes;
      delete this.fields.barcode;
      return;
    }

    if (!Array.isArray(v))
      throw new TypeError(`barcodes must be an Array, received ${typeof v}`);

    // Barcodes dictionary: https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/LowerLevel.html#//apple_ref/doc/uid/TP40012026-CH3-SW3
    for (const barcode of v) {
      if (!BARCODES_FORMAT.has(barcode.format))
        throw new TypeError(
          `Barcode format value ${barcode.format} is invalid!`,
        );
      if (typeof barcode.message !== 'string')
        throw new TypeError('Barcode message string is required');
      if (typeof barcode.messageEncoding !== 'string')
        throw new TypeError('Barcode messageEncoding is required');
    }

    // copy array
    this.fields.barcodes = [...v];
  }

  /**
   * Adds a location where a pass is relevant.
   *
   * @param {number[] | { lat: number, lng: number, alt?: number } | { longitude: number, latitude: number, altitude?: number }} point
   * @param {string} [relevantText]
   * @returns {this}
   */
  addLocation(
    point:
      | number[]
      | { lat: number; lng: number; alt?: number }
      | { longitude: number; latitude: number; altitude?: number },
    relevantText?: string,
  ): this {
    const { longitude, latitude, altitude } = getGeoPoint(point);
    const location: import('../interfaces').Location = {
      longitude,
      latitude,
    };
    if (altitude) location.altitude = altitude;
    if (typeof relevantText === 'string') location.relevantText = relevantText;
    if (!Array.isArray(this.fields.locations))
      this.fields.locations = [location];
    else this.fields.locations.push(location);
    return this;
  }

  get locations(): ApplePass['locations'] {
    return this.fields.locations;
  }

  set locations(v: ApplePass['locations']) {
    delete this.fields.locations;
    if (!v) return;
    if (!Array.isArray(v)) throw new TypeError(`locations must be an array`);
    else for (const location of v) this.addLocation(location);
  }
}
