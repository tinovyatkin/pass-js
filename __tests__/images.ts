'use strict';

import * as path from 'path';

import { PassImages } from '../src/lib/images';

describe('PassImages', () => {
  it('has class properties', () => {
    const img = new PassImages();
    expect(img.loadFromDirectory).toBeInstanceOf(Function);
  });

  it('reads all images from directory without localized images', async () => {
    const img = new PassImages();
    const imgDir = path.resolve(__dirname, '../images/');
    await img.loadFromDirectory(imgDir);
    expect(img.count).toBe(18);
  });

  it('should read localized images', async () => {
    const img = new PassImages();
    const imgDir = path.resolve(__dirname, './resources/passes/Generic');
    await img.loadFromDirectory(imgDir);
    expect(img.count).toBe(5);
    const arr = await img.toArray();
    expect(arr).toBeInstanceOf(Array);
    expect(arr.map(f => f.path)).toMatchSnapshot();
  });
});
