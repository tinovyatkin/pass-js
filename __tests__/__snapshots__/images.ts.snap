// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PassImages should read localized images 1`] = `
Array [
  "icon.png",
  "ru.lproj/logo.png",
  "ru.lproj/logo@2x.png",
  "zh_CN.lproj/logo.png",
  "zh_CN.lproj/logo@2x.png",
]
`;
